File: Engine/iOS/RecorderCore.swift
//
//  RecorderCore.swift
//  DespiaScript
//
//  The shared `recorder` module core: the metering curve and the state machine, including
//  how an interruption moves through it. The law is the corpus,
//  OpenSource/Conformance/recorder/ (parity/F14-recorder.md). The twin of :core
//  RecorderCore.kt and of the web @despia-native/kernel recorder-core.ts.
//
//  WHY THESE TWO PARTS AND NOT THE RECORDING. Capturing audio is entirely platform work
//  (AVAudioRecorder, AVAudioSession) and belongs in the module facet. What cannot live
//  there is the METER CURVE, because this platform reports average power in dBFS and
//  Android reports a 16-bit linear amplitude, so without one shared fold the same voice
//  fills the bar on one platform and barely moves it on the other. And the STATE MACHINE,
//  because interruption handling is where every naive recorder loses a take.
//

import Foundation

public enum RecorderCore {

    /// 0.0 on the meter. Quieter than this is silence as far as a level meter is concerned.
    public static let floorDb: Double = -60

    /// The meter tick. Ten a second looks continuous and costs nothing per event.
    public static let meterIntervalMs = 100

    /// `interrupted` is a STATE rather than an error because a phone call is a normal thing
    /// that happens during a recording.
    public static let states: [String] = ["idle", "recording", "paused", "interrupted"]

    /// m4a (AAC) is the default and the only one both mobile platforms encode natively.
    public static let formats: [String] = ["m4a", "wav", "opus"]

    /// Fold an author's format spelling; nil is `unsupported_format`.
    public static func foldFormat(_ name: String?) -> String? {
        var key = (name ?? "").trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if key.hasPrefix(".") { key.removeFirst() }
        if key.isEmpty { return "m4a" }
        switch key {
        case "mp4", "aac": return "m4a"
        case "wave": return "wav"
        case "ogg": return "opus"
        default: return formats.contains(key) ? key : nil
        }
    }

    /// Four decimals, half AWAY FROM ZERO. Spelled out rather than left to each language's
    /// default, because the JS and JVM rounding rules disagree on negatives and dB values
    /// are negative.
    private static func round4(_ value: Double) -> Double {
        let scaled = (abs(value) * 10000).rounded() / 10000
        return value < 0 ? -scaled : scaled
    }

    /// dBFS to a 0...1 meter level, LINEAR IN DECIBELS. -60 dBFS or quieter is 0, 0 dBFS is
    /// 1, and -30 dBFS sits at exactly half, which is what makes a bar look like the
    /// loudness a person hears rather than like the raw amplitude.
    public static func meterLevel(_ db: Double) -> Double {
        guard db.isFinite else { return 0 }
        if db <= floorDb { return 0 }
        if db >= 0 { return 1 }
        return round4((db - floorDb) / (0 - floorDb))
    }

    /// A linear amplitude reading to dBFS. Android's `getMaxAmplitude()` reports a 16-bit
    /// peak (full scale 32767) and the web's AnalyserNode reports a normalised float (full
    /// scale 1); one conversion serves both.
    public static func amplitudeDb(_ amplitude: Double, fullScale: Double) -> Double {
        guard amplitude.isFinite, fullScale.isFinite, fullScale > 0 else { return floorDb }
        let ratio = abs(amplitude) / fullScale
        guard ratio > 0 else { return floorDb }
        let db = 20 * log10(ratio)
        if db <= floorDb { return floorDb }
        return db >= 0 ? 0 : round4(db)
    }

    /// The amplitude path in one step: amplitude in, meter level out.
    public static func meterFromAmplitude(_ amplitude: Double, fullScale: Double) -> Double {
        meterLevel(amplitudeDb(amplitude, fullScale: fullScale))
    }

    /// What a transition produced: the new state, an optional broadcast the module must
    /// emit, and an optional refusal code.
    public struct Transition {
        public let state: String
        public let emit: String?
        public let error: String?
        public init(_ state: String, emit: String? = nil, error: String? = nil) {
            self.state = state
            self.emit = emit
            self.error = error
        }
    }

    /// The state machine. One table, three renderers, and the interruption path is in it
    /// rather than in each platform's notification handler, which is where it usually rots.
    ///
    /// Events: start, pause, resume, stop, cancel, interrupt, endInterruption, maxDuration.
    ///
    /// The two rules worth stating out loud: a second `start` is `busy`, never a second
    /// file; and a returning session moves to `paused`, never straight back to `recording`,
    /// because coming back from a phone call to find the app quietly recording again is
    /// worse than a take that needs one tap.
    public static func transition(_ state: String, _ event: String) -> Transition {
        let from = states.contains(state) ? state : "idle"
        switch event {
        case "start":
            return from == "idle" ? Transition("recording") : Transition(from, error: "busy")
        case "pause":
            if from == "recording" || from == "paused" { return Transition("paused") }
            return Transition(from, error: "not_recording")
        case "resume":
            if from == "paused" || from == "interrupted" || from == "recording" { return Transition("recording") }
            return Transition(from, error: "not_recording")
        case "stop", "cancel":
            return from == "idle" ? Transition("idle", error: "not_recording") : Transition("idle")
        case "interrupt":
            return from == "recording" ? Transition("interrupted", emit: "interrupted") : Transition(from)
        case "endInterruption":
            return from == "interrupted" ? Transition("paused", emit: "resumable") : Transition(from)
        case "maxDuration":
            return from == "recording" ? Transition("idle") : Transition(from)
        default:
            return Transition(from, error: "not_recording")
        }
    }
}

File: Engine/Android/core/src/main/kotlin/despia/engine/RecorderCore.kt
//
//  RecorderCore.kt - the shared `recorder` module core (:core, pure JVM): the metering
//  curve and the state machine, including how an interruption moves through it. The law is
//  the corpus, OpenSource/Conformance/recorder/ (parity/F14-recorder.md). The twin of the
//  web @despia-native/kernel recorder-core.ts and of Swift RecorderCore.
//
//  WHY THESE TWO PARTS AND NOT THE RECORDING. Capturing audio is entirely platform work
//  (MediaRecorder, AudioRecord, MediaCodec) and belongs in the module facet. What cannot
//  live there is the METER CURVE, because iOS reports average power in dBFS and Android
//  reports a 16-bit linear amplitude, so without one shared fold the same voice fills the
//  bar on one platform and barely moves it on the other. And the STATE MACHINE, because
//  interruption handling is where every naive recorder loses a take.
//
package despia.engine

import kotlin.math.abs
import kotlin.math.log10
import kotlin.math.roundToLong

object RecorderCore {

    /** 0.0 on the meter. Quieter than this is silence as far as a level meter is concerned. */
    const val FLOOR_DB: Double = -60.0

    /** The meter tick. Ten a second looks continuous and costs nothing per event. */
    const val METER_INTERVAL_MS: Int = 100

    /** `interrupted` is a STATE rather than an error because a phone call is a normal thing
     *  that happens during a recording. */
    val STATES: List<String> = listOf("idle", "recording", "paused", "interrupted")

    /** m4a (AAC) is the default and the only one both mobile platforms encode natively. */
    val FORMATS: List<String> = listOf("m4a", "wav", "opus")

    /** Fold an author's format spelling; null is `unsupported_format`. */
    fun foldFormat(name: String?): String? {
        val key = (name ?: "").trim().lowercase().removePrefix(".")
        if (key.isEmpty()) return "m4a"
        return when (key) {
            "mp4", "aac" -> "m4a"
            "wave" -> "wav"
            "ogg" -> "opus"
            else -> if (FORMATS.contains(key)) key else null
        }
    }

    /** Four decimals, half AWAY FROM ZERO. Spelled out rather than left to each language's
     *  default, because the JS and JVM rounding rules disagree on negatives and dB values
     *  are negative. */
    private fun round4(value: Double): Double {
        val scaled = (abs(value) * 10000).roundToLong() / 10000.0
        return if (value < 0) -scaled else scaled
    }

    /**
     * dBFS to a 0..1 meter level, LINEAR IN DECIBELS. -60 dBFS or quieter is 0, 0 dBFS is 1,
     * and -30 dBFS sits at exactly half, which is what makes a bar look like the loudness a
     * person hears rather than like the raw amplitude.
     */
    fun meterLevel(db: Double): Double {
        if (db.isNaN() || db.isInfinite()) return 0.0
        if (db <= FLOOR_DB) return 0.0
        if (db >= 0.0) return 1.0
        return round4((db - FLOOR_DB) / (0.0 - FLOOR_DB))
    }

    /**
     * A linear amplitude reading to dBFS. `MediaRecorder.getMaxAmplitude()` reports a 16-bit
     * peak (full scale 32767) and the web's AnalyserNode reports a normalised float (full
     * scale 1); one conversion serves both.
     */
    fun amplitudeDb(amplitude: Double, fullScale: Double): Double {
        if (amplitude.isNaN() || fullScale.isNaN() || fullScale <= 0.0) return FLOOR_DB
        val ratio = abs(amplitude) / fullScale
        if (ratio <= 0.0) return FLOOR_DB
        val db = 20.0 * log10(ratio)
        if (db <= FLOOR_DB) return FLOOR_DB
        return if (db >= 0.0) 0.0 else round4(db)
    }

    /** The Android path in one step: amplitude in, meter level out. */
    fun meterFromAmplitude(amplitude: Double, fullScale: Double): Double =
        meterLevel(amplitudeDb(amplitude, fullScale))

    /** What a transition produced: the new state, an optional broadcast the module must
     *  emit, and an optional refusal code. */
    data class Transition(val state: String, val emit: String? = null, val error: String? = null)

    /**
     * The state machine. One table, three renderers, and the interruption path is in it
     * rather than in each platform's focus-change callback, which is where it usually rots.
     *
     * Events: start, pause, resume, stop, cancel, interrupt, endInterruption, maxDuration.
     *
     * The two rules worth stating out loud: a second `start` is `busy`, never a second file;
     * and a returning session moves to `paused`, never straight back to `recording`, because
     * coming back from a phone call to find the app quietly recording again is worse than a
     * take that needs one tap.
     */
    fun transition(state: String, event: String): Transition {
        val from = if (STATES.contains(state)) state else "idle"
        return when (event) {
            "start" -> if (from == "idle") Transition("recording") else Transition(from, error = "busy")
            "pause" -> when (from) {
                "recording", "paused" -> Transition("paused")
                else -> Transition(from, error = "not_recording")
            }
            "resume" -> when (from) {
                "paused", "interrupted", "recording" -> Transition("recording")
                else -> Transition(from, error = "not_recording")
            }
            "stop", "cancel" ->
                if (from == "idle") Transition("idle", error = "not_recording") else Transition("idle")
            "interrupt" ->
                if (from == "recording") Transition("interrupted", emit = "interrupted") else Transition(from)
            "endInterruption" ->
                if (from == "interrupted") Transition("paused", emit = "resumable") else Transition(from)
            "maxDuration" -> if (from == "recording") Transition("idle") else Transition(from)
            else -> Transition(from, error = "not_recording")
        }
    }
}
